quil-rs/src/stats/mod.rs

```diff
@@ -0,0 +1,270 @@
+mod execution_graph;
+
+#[cfg(test)]
+pub(crate) mod test_programs;
+
+// Use quil_rs::program::graph for pulse-level programs, but not qubit programs.
+
+use std::collections::HashSet;
+use std::ops::Neg;
+
+use crate::{
+    instruction::{Instruction, Qubit},
+    Program,
+};
+use execution_graph::{Error as ExecutionGraphError, ExecutionGraph};
+
+pub trait InstructionsSource {
+    fn body_instructions(&self) -> impl Iterator<Item = &Instruction> + '_;
+    fn num_instructions(&self) -> usize;
+    fn get_used_qubits(&self) -> impl IntoIterator<Item = &Qubit>;
+}
```

Comment:
`impl Trait` as the return type in traits is a very recent thing, IIRC. we might consider using associated types (i.e. `type IntoInstructionIter = &[Instruction]`) for better backwards compatibility.